pkg/controller/cainjector/sources.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cainjector
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+)
+
+var (
+	// WantInjectAnnotation is the annotation that specifies that a particular
+	// object wants injection of CAs.  It takes the form of a reference to a certificate
+	// as namespace/name.  The certificate is expected to have the is-serving-for annotations.
+	WantInjectAnnotation = "certmanager.k8s.io/inject-ca-from"
+
+	// WantInjectFromSecretAnnotation is the annotation that specifies that a particular
+	// object wants injection of CAs.  It takes the form of a reference to a Secret
+	// as namespace/name.
+	WantInjectFromSecretAnnotation = "certmanager.k8s.io/inject-ca-from-secret"
+
+	// WantInjectAPIServerCAAnnotation, if set to "true", will make the cainjector
+	// inject the CA certificate for the Kubernetes apiserver into the resource.
+	// It discovers the apiserver's CA by inspecting the service account credentials
+	// mounted into the
+	WantInjectAPIServerCAAnnotation = "certmanager.k8s.io/inject-apiserver-ca"
+)
+
+// caDataSource knows how to extract CA data given a provided InjectTarget.
+// This allows adaptable implementations of fetching CA data based on
+// configuration given on the injection target (e.g. annotations).
+
+type caDataSource interface {
+	// Configured returns true if this data source should be used for the given
+	// InjectTarget, i.e. if it has appropriate annotations enabled to use the
+	// annotations.
+	Configured(log logr.Logger, metaObj metav1.Object) bool
+
+	// ReadCA reads the CA that should be injected into the InjectTarget based
+	// on the configuration provided in the InjectTarget.
+	// ReadCA may return nil, nil if the CA data cannot be read.
```

Comment:
nil, nil -> nil

Reply:
nil, nil is correct - ReadCA returns two args and in some cases, both are nil 😄